File: Hashing_and_HashMap_problems/JewelsAndStones.java

// Problem link : https://leetcode.com/problems/jewels-and-stones/description/?envType=problem-list-v2&envId=string&status=TO_DO&difficulty=EASY

package Hashing_and_HashMap_problems;

import java.util.HashSet;

public class JewelsAndStones {
    public static int numJewelsInStones1(String jewels, String stones) {
        // convert jewels string to a Set for efficient lookup.
        HashSet<Character> jewelSet = new HashSet<>();
        for(char ch : jewels.toCharArray()){
            jewelSet.add(ch);
        }

        // initialize a counter to track the number of jewels in stones.
        int count = 0;

        // loop through each character in stones.
        for(char stone : stones.toCharArray()){
            // check if the current stone is in the jewelSet.
            if(jewelSet.contains(stone)){
                // increment the count if it is a jewel
                count++;
            }
        }

        // return the total count of jewels in stones.
        return count;
    }

    public static int numJewelsInStones(String jewels, String stones) {
        // initialize a counter to track the number of jewels in stones.
        int count = 0;

        // loop through each character in stones.
        for(char stone : stones.toCharArray()){
            // check if the current stone is in the jewel.
            if(jewels.indexOf(stone) != -1){
                // increment the count if it is a jewel
                count++;
            }
        }

        // return the total count of jewels in stones.
        return count;
    }

    public static void main(String[] args) {
        String jewels = "aA", stones = "aAAbbbb";

        // String jewels = "z", stones = "ZZ";
        System.out.println(numJewelsInStones(jewels, stones));
    }
}

File: Hashing_and_HashMap_problems/JewelsAndStones.js

// Problem link : https://leetcode.com/problems/jewels-and-stones/description/?envType=problem-list-v2&envId=string&status=TO_DO&difficulty=EASY

let numJewelsInStones1 = function(jewels, stones) {
    let frqMap = {};

    for(let i=0; i<stones.length; i++){
        if(frqMap[stones[i]]){
            frqMap[stones[i]]++;
        }else{
            frqMap[stones[i]] = 1;
        }
    }

    let count = 0;
    for(let i=0; i<jewels.length; i++){
        if(frqMap[jewels[i]]){
            count += frqMap[jewels[i]];
        }
    }

    return count;;
};

let numJewelsInStones = function(jewels, stones) {
    // convert jewels string to a Set for efficient lookup.
    let jewelsSet = new Set(jewels);
    
    // initialize a counter to track the number of jewels in stones.
    let count = 0;

    // loop through each character in stones.
    for(let s of stones){
        // check if the current stone is in the jewelSet.
        if(jewelsSet.has(s)){
            // increment the count if it is a jewel
            count++;
        }
    }

    // return the total count of jewels in stones.
    return count;
};

let jewels = "aA", stones = "aAAbbbb";

// let jewels = "z", stones = "ZZ";
console.log(numJewelsInStones(jewels, stones));
